### PR TITLE
Implement GetAttrVariable.as_python_constant()

### PR DESCRIPTION
**Pull request by rec**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134216

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_dynamo/variables/misc.py [hunk 1]
```diff
@@ -946,6 +946,13 @@ def create_getattr_proxy(base_proxy: torch.fx.Proxy, attr):
     def as_proxy(self):
         return GetAttrVariable.create_getattr_proxy(self.obj.as_proxy(), self.name)
 
+    def as_python_constant(self):
+        constant = self.obj.as_python_constant()
+        try:
+            return getattr(constant, self.name)
+        except AttributeError:
+            raise NotImplementedError from None
```
Comment (amjames) [inline torch/_dynamo/variables/misc.py:954]: Nit: 
```suggestion
            raise NotImplementedError(f"{self} is not a constant") from None
```

**Review by amjames — APPROVED** [review]

**Review by williamwen42 — COMMENTED** [review]
Add a test to `test/dynamo/test_misc.py` as well?

**Comment by rec** [comment]
> Add a test to `test/dynamo/test_misc.py` as well?

Hey, William, pleased to meet you, and thanks for taking a look at this pull request!

Aside from the 10 failing tests that now pass, I had some trouble creating a new test just for this functionality.

After all my attempts failed, I ran all the dynamo tests to see which ones actually ended up calling this new method.

Only one did, and it's a fairly complicated test involving random number generation.  I copied that test and tried to make it as small as possible - you can see the original and the smaller version [here](https://github.com/rec/pytorch/blob/968db063e0866b5b688ff9f6588ed19933173b2f/test/dynamo/test_unspec.py#L224-L280).

I could easily use that slightly smaller test to prove that this method is being successfully called, but it isn't very satisfying or explanatory to the reader.

Should I do that?

Or do you have some idea for a much clearer test?

Or could I just point to the new tests that now pass as evidence?

**Comment by williamwen42** [comment]
I was hoping we could just get a minimal repro from the previously failing tests or construct a situation where we're trying to turn a getattr variable into a constant, which before this PR would result in a failure.

If this is hard to do, then having a test that just goes through this new codepath is still helpful since we can make sure we don't regress on this behavior in the future.

**Comment by rec** [comment]
@williamwen42: I wrote a test which does exercise and check it.

The compiled function is irreducible and not too long but comes with no explanation as to why it actually exercises that code when slightly shorter tests do not other. 

#131354, which is stacked on this, should run `GetAttrVariable.as_python_constant()` "more often", as dynamo will have a better idea of what is constant and not.

File: test/dynamo/test_misc.py [hunk 2]
```diff
@@ -11333,6 +11333,40 @@ def fn(x):
...
+            random.Random().setstate(rand1.getstate())
+            return x + rand1.random()
+
+        def get_rng(x):
+            rand1 = random.Random(1)
+            orig_random = rand1.random
+            rand1.random = lambda: orig_random()
+            return rand1
+
+        as_python_constant = GetAttrVariable.as_python_constant
```
Comment (williamwen42) [inline test/dynamo/test_misc.py:11350]: Can we use something like `unittest.mock` to patch and detect calls instead of manually patching?

Comment (rec) [reply test/dynamo/test_misc.py]: @williamwen42 

Yes, of course, silly me!

It turns out that `unitest.mock.patch.object` fits the job very nicely, take a look.

**Review by williamwen42 — APPROVED** [review]

**Pull request closed without merging**